### PR TITLE
librbd: add missing shutdown states to managed lock helper

### PR DESCRIPTION
**Pull request by dillaman**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman observing rbd-mirror crashes when running it on teuthology [1]. I am not sure it is related though, I am running it on mira (which I did not do for long time) because smithi overloaded these days, so it might be an old problem just exposed in mira. It crashes in `:LeaderWatcher::shut_down` here is an example:
```
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz:2019-02-21 10:08:19.841 7fc225d3b700  5 rbd::mirror::PoolWatcher: 0x7fc21400e1c0 handle_rewatch_complete: r=-108
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz:2019-02-21 10:08:19.841 7fc225d3b700  0 rbd::mirror::PoolWatcher: 0x7fc21400e1c0 handle_rewatch_complete: detected client is blacklisted
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz:2019-02-21 10:08:19.841 7fc225d3b700 10 librbd::watcher::RewatchRequest: 0x7fc214014190 rewatch 
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz:2019-02-21 10:08:19.841 7fc237afbac0 -1 *** Caught signal (Aborted) **
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: in thread 7fc237afbac0 thread_name:rbd-mirror
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz:
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: ceph version 14.0.1-3913-g6945440 (69454404fc5d1954419663246c470d3326916bc9) nautilus (dev)
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: 1: (()+0x12890) [0x7fc22e2ba890]
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: 2: (pthread_cond_wait()+0x243) [0x7fc22e2b59f3]
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: 3: (rbd::mirror::LeaderWatcher<librbd::ImageCtx>::shut_down()+0x19d) [0x564816df9c7d]
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: 4: (rbd::mirror::PoolReplayer<librbd::ImageCtx>::shut_down()+0x77) [0x564816da1b37]
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: 5: (rbd::mirror::Mirror::update_pool_replayers(std::map<long, std::set<rbd::mirror::PeerSpec, rbd::mirror::ClusterWatcher::PeerSpecCompare, std::
allocator<rbd::mirror::PeerSpec> >, std::less<long>, std::allocator<std::pair<long const, std::set<rbd::mirror::PeerSpec, rbd::mirror::ClusterWatcher::PeerSpecCompare, std::allocator<rbd::mirror::PeerSpec> > > > > const&)+0x8ea) [0x564816d9ac7a]
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: 6: (rbd::mirror::Mirror::run()+0xa6) [0x564816d9b9f6]
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: 7: (main()+0x3b6) [0x564816d8bd86]
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: 8: (__libc_start_main()+0xe7) [0x7fc22d526b97]
3621802/remote/mira058/log/cluster1-client.mirror.1.12411.log.gz: 9: (_start()+0x2a) [0x564816d9569a]
```
There was also a crash in `PerfCounters::inc`:
```
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: ceph version 14.0.1-3913-g6945440 (69454404fc5d1954419663246c470d3326916bc9) nautilus (dev)
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 1: (()+0x11390) [0x7f1dc0004390]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 2: (gsignal()+0x38) [0x7f1dbf852428]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 3: (abort()+0x16a) [0x7f1dbf85402a]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 4: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1a3) [0x7f1dc04ef633]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 5: (ceph::__ceph_assertf_fail(char const*, char const*, int, char const*, char const*, ...)+0) [0x7f1dc04ef7bd]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 6: (PerfCounters::inc(int, unsigned long)+0x79) [0x7f1dc06e5c59]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 7: (rbd::mirror::ImageReplayer<librbd::ImageCtx>::handle_process_entry_safe(journal::ReplayEntry const&, utime_t const&, int)+0x10a) [0x663aca]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 8: (Context::complete(int)+0x9) [0x5d0019]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 9: (ThreadPool::worker(ThreadPool::WorkThread*)+0x9d6) [0x7f1dc0593db6]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 10: (ThreadPool::WorkThread::entry()+0x10) [0x7f1dc0595000]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 11: (()+0x76ba) [0x7f1dbfffa6ba]
3621814/remote/mira010/log/cluster1-client.mirror.0.13541.log.gz: 12: (clone()+0x6d) [0x7f1dbf92441d]
```
This probably related to the recently added replication counters functionality.

[1] http://pulpito.ceph.com/trociny-2019-02-21_07:37:09-rbd-wip-mgolub-testing-distro-basic-mira/

**Comment by dillaman** [comment]
@trociny I'll take a look at both. The first one is what I was trying to fix with this PR since I have teuthology configured to send an abort signal to ```rbd-mirror``` if it fails to shut down after several minutes (so that I can analyze the core).

**Comment by dillaman** [comment]
The ```PerfCounters``` issue is because the ```g_perf_counters``` is deleted before everything is shut down. I'll fix it as part of this PR.

**Comment by dillaman** [comment]
jenkins test docs

**Comment by dillaman** [comment]
I am going to run this through teuthology to see if any additional deadlocks appear

**Comment by dillaman** [comment]
Only remaining failure is a non-shutdown issue [1][2]

[1] http://pulpito.ceph.com/jdillaman-2019-02-23_10:41:10-rbd-wip-jd-testing-distro-basic-smithi/
[2] http://tracker.ceph.com/issues/38463

File: src/tools/rbd_mirror/ImageReplayer.cc [hunk 1]
```diff
@@ -501,15 +501,19 @@ void ImageReplayer<I>::bootstrap() {
...
-    &m_local_image_ctx, m_local_image_id, m_remote_image.image_id,
-    m_global_image_id, m_local_mirror_uuid, m_remote_image.mirror_uuid,
-    m_remote_journaler, &m_client_state, &m_client_meta, ctx,
-    &m_resync_requested, &m_progress_cxt);
-
+  BootstrapRequest<I> *request = nullptr;
   {
     Mutex::Locker locker(m_lock);
+    if (on_start_interrupted(m_lock)) {
+      return;
```
Comment (trociny) [inline src/tools/rbd_mirror/ImageReplayer.cc]: Need to move `ctx` creation below this return. I suppose this is the cause of the observed valgrind error:

 http://qa-proxy.ceph.com/teuthology/trociny-2019-02-25_13:52:38-rbd-wip-mgolub-testing-distro-basic-smithi/3636974/teuthology.log

Comment (dillaman) [reply src/tools/rbd_mirror/ImageReplayer.cc]: Indeed -- thx. Fix pushed.

**Comment by trociny** [comment]
@dillaman and there were two cases in the last run [1] when it still aborted in the leader watcher shutdown (see the failed stress tests).

[1] http://pulpito.ceph.com/trociny-2019-02-25_13:52:38-rbd-wip-mgolub-testing-distro-basic-smithi/

**Comment by dillaman** [comment]
@trociny Two different issues fixed: (1) trash watcher got stuck in a ```-ENOENT``` loop when attempting to create the object in a deleted pool and (2) image replayer bootstrap failed to stop when manually stopped since it ```on_start_interrupted``` was checking the wrong variable.

**Pull request merged**